### PR TITLE
Reconcile with upstream wesleytodd/express-openapi

### PR DESCRIPTION
**Pull request by enov**
## About the changes
Reconcile with upstream: @wesleytodd made some changes 2 weeks ago that are worth adding here. Reconciling the projects sooner would help us resolve conflicts before they get diverged to the point of no return.

This reconcilation effort has been made with @SWheeler17 (aka @Megapixel99).

Tests passed when we ran them locally.

## Conflicts:
- .github/workflows/test.yml
- lib/validate.js
- package.json

## Conflict Resolution:
- Remove Nodejs v12.x from Github workflow test
- Remove `extendRefs` option, as it's not needed for ajv 8+. See #9
- Use 'url.URL' constructor instead of the deprecated ([?](https://nodejs.org/api/url.html#urlparseurlstring-parsequerystring-slashesdenotehost)) url.parse
- Appease eslint with some minor whitespace changes

### CODE REVIEW

**Comment by thomasheartman** [comment]
@enov Thanks for the PR! I don't have time to go through it right now, but I'll have a look at it next week 😄 

In the meantime: You mention that there are
> some changes [...] that are worth adding here

Could you give a rough outline of what those changes are? Is it the yaml endpoint? The removal of the deprecated url.parse? Something else?

**Comment by Megapixel99** [comment]
The changes we made, aside from what was already stated in conflict resolution, is:
 - [The addition of the yaml endpoint](https://github.com/Unleash/express-openapi/pull/12/files#diff-e727e4bdf3657fd1d798edcd6b099d6e092f8573cba266154583a746bba0f346R137)
 - [The addition of a comment telling standard to ignore no-case-declarations for the next line](https://github.com/Unleash/express-openapi/pull/12/files#diff-cabded08ff9373812a499e53913ccaca1d50f6d981eefa5c45f8cb88809a9cf7R53)
 - [The addition of the `strict` option and removal of `extendedRefs` from AJV](https://github.com/Unleash/express-openapi/pull/12/files#diff-cabded08ff9373812a499e53913ccaca1d50f6d981eefa5c45f8cb88809a9cf7R93)
   - when we were doing research we came to the belief that, per https://github.com/ajv-validator/ajv/issues/260, `extendedRefs` should have been `$refs`. After looking at the data being passed into `extendedRefs` while running the test; however, one of the two times it is used in the test, it is an empty object and the second time the object is set to `{ coerce: true }` (implemented here: https://github.com/ajv-validator/ajv/pull/246). The removal of `extendedRefs` did not negatively affect the tests when run locally and the inclusion of `extendedRefs` resulted in the error: `NOT SUPPORTED: option extendRefs. Deprecated ignoreKeywordsWithRef can be used instead.`
 - [replaced the deprecated `dataPath` with `instancePath`](https://github.com/Unleash/express-openapi/pull/12/files#diff-b7f49e7d6717fcd3fc4e6864e9956dbcafc8d00362412f379014ea88e2743ec1L95) which is mentioned in multiple PRs across Github such as: https://github.com/ajv-validator/ajv/pull/1487
 - Updated various dependencies such as `swagger-ui-dist`, `swagger-parser`, and `ajv` with the version of the packages in the weslytodd version

File: lib/validate.js
Comment (thomasheartman) [inline lib/validate.js:53]: Why are you ignoring `no-case-declarations` for the next line here? Looking at the [eslint docs for this lint](https://eslint.org/docs/latest/rules/no-case-declarations), it doesn't seem to pertain to anything here?

Unless I'm misunderstanding something here, I'd suggest to remove this line.

```suggestion
```

File: lib/validate.js
Comment (thomasheartman) [inline lib/validate.js:93]: Minor comment, but is this just converting a boolean or undefined/null into a boolean or are there other edge cases? (I guess if this is JS, then you can realistically pass literally anything)

I was going to suggest using the boolean constructor, but that might have unintended consequences if you pass, say, the number 5.

**Review by thomasheartman — COMMENTED** [review]
Thanks for the detailed overview! 😄 I've looked through it now, and it looks like a solid update to me 👏🏼  I have a few comments, however, that I'd love to hear your thoughts on before getting this in. I'll also go ahead and read up a bit on the extendedRefs removal. I think it should be alright, but just wanna make doubly sure.

File: lib/validate.js
Comment (thomasheartman) [reply lib/validate.js:93]: ~~Oh, and also: what is this option used for? I don't see it referenced anywhere in the PR. What does it do?~~

Never mind, it gets passed to AJV, of course. Sorry, my bad 😅

File: lib/validate.js
Comment (enov) [reply lib/validate.js:53]: Quoting the eslint docs:
> This rule disallows lexical declarations (let, const, function and class) in case/default clauses. The reason is that the lexical declaration is visible in the entire switch block but it only gets initialized when it is assigned, which will only happen if the case where it is defined is reached.
>
> To ensure that the lexical declaration only applies to the current case clause wrap your clauses in blocks.

There is a constant declaration on L54. To ensure that the declaration does not leak to out of the case block, I would personally wrapped the case block with braces, like this:

```
case 'header': {
  const name = p.name.toLowerCase()
  reqSchema.properties.headers.properties[name] = p.schema
  p.required && !reqSchema.properties.headers.required.includes(p.name) && reqSchema.properties.headers.required.push(name)
  break
}
```
However, the [author chose to ignore the rule](https://github.com/Unleash/express-openapi/commit/98add2cbd89818cd90d43aa0d6cc8d6733dca1fe#diff-cabded08ff9373812a499e53913ccaca1d50f6d981eefa5c45f8cb88809a9cf7R53).

Comment (wesleytodd) [reply lib/validate.js:53]: Feel free to open a pr if y'all want. As with any opinionated lint rules, they are there to protect against making mistakes without realizing it. I for sure think moving it out and using `let` is worse since it would separate it from the context in which it was used. Linting would also catch it leaking, so I don't think it matters, but would totally accept a PR adding the `{}` since that is more explicit about the intent anyway.

Comment (enov) [reply lib/validate.js:53]: No offense @wesleytodd, if I sounded bad :pray: and thank you  for chiming in. I appreciate your hard work.
And yes, I don't think it matters either; that's why I left it as is. Once this is merged, I'll do a pull request to both repos.
Cheers!

Comment (wesleytodd) [reply lib/validate.js:53]: Oh I didn't take offense. It just code, lots of ways to approach problems. I just figured an explanation of the thinking behind it might help. And I would love a PR with stuff y'all have found using the lib!

Comment (thomasheartman) [reply lib/validate.js:53]: Sounds great! And yeah, sorry, my mind must've been somewhere else when I was looking through this. I didn't register that the change was in a switch block at all. Yeah, given that, it's totally fine 😅

**Review by thomasheartman — APPROVED** [review]
Thanks again for taking the time to update the code and making the PR here 🙏🏼 I think this is good to go 👏🏼  Appreciate your patience 🙇🏼

**Comment by enov** [comment]
@thomasheartman Thank you for merging the pull request, I appreciate it.

This being a downstream fork of `wesleytodd/express-openapi`, a better merge strategy would have been to merge with a merge commit without squashing all commits into a single commit. At this point, while the code base has been reconciled with the upstream, the git commits are not. We have one more commit ahead of `wesleytodd:main` than we used to have.

That being said, could you please publish a new version to npm. Thanks again!

**Comment by thomasheartman** [comment]
> This being a downstream fork of wesleytodd/express-openapi, a better merge strategy would have been to merge with a merge commit without squashing all commits into a single commit

Ah, that's a good point. We **could** revert and merge the merge commit instead 🤔 

> That being said, could you please publish a new version to npm. Thanks again!

I was going to, but wanted to make sure it worked for Unleash first, but had some unexpected issues. I've been preoccupied this week and will be out next week, but I will try to see if I can get to it after that. 

Out of curiosity, do you depend on this package in any way? If so, why not depend on the upstream?

**Comment by enov** [comment]
> We could revert and merge the merge commit instead

That's a possible solution. Then the branch would be 32 commits ahead (add a "revert" commit, plus a new "merge" commit to the existing count of "30 commits ahead") and 0 commits behind `wesleytodd:main`.

> you depend on this package in any way? If so, why not depend on the upstream?

Correct, we have a dozen of APIs exposed using this package. At that time (a year ago?), we were looking for replacement of [Exegisis](https://github.com/exegesis-js/exegesis) and [Apidoc](https://github.com/apidoc/apidoc) (nothing wrong with those, except that our needs were simpler). We found Wes Todd's library which seemed to us kind of abandoned at that time. Your fork looked better maintained. Plus, we did not need redoc... Now, our k8s ingress is configured to have `swagger-ui-init.js` and Co bypass our Auth policy, so switching back needs some work.

Thank you for time looking at this.

**Comment by thomasheartman** [comment]
Hey! Apologies for the delay on this, but we'll prioritize this a task next week. Appreciate your patience.

**Comment by thomasheartman** [comment]
Hey, @enov! I've spent some time looking at this and discussing with other members of the team today. I think what we've decided is that we'd like to dismantle this fork if possible (to lessen the maintenance burden). It was first created by someone who isn't around anymore, and none of us can find a compelling reason to keep the fork open. I'm currently in the process of talking to Wesley T about upstreaming the one little change we made.

So unless we can find a good reason not to go with the upstream, it probably means that we're not going to be releasing an updated package, and we'll likely archive this in not too long. Apologies for the let-down there. But that said, it's all open, so you should be able to just pick up your own fork 💁🏼  Still, though: very much appreciate the effort you put into updating this one 🙏🏼

**Comment by enov** [comment]
Thank you @thomasheartman! Wes added me as a maintainer to his repo, so there is no need to fork. Dismantling / archiving this fork is best: we can all collaborate on a single repo. Also, we noticed that [OpenAPI formats will be ignored](https://github.com/wesleytodd/express-openapi/pull/42). That is still missing here. So, it's good not to (or never) publish a new version at this stage, unless you want me to look at that update this fork again, just one more time; I don't mind.

**Comment by thomasheartman** [comment]
Sounds great! We'll work towards that, then 😄 Thanks for your input!

**Pull request merged**
